### PR TITLE
feature: ssl_certby_lua* support for stream_lua.

### PR DESCRIPTION
**Pull request by dndx**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-core project.

### CODE REVIEW

**Comment by agentzh** [comment]
@thibaultcha Will you have a look at this? If everything is OK, please merge it :) Thanks!

File: lib/ngx/ssl.lua [hunk 1]
```diff
@@ -17,53 +17,149 @@ local get_string_buf = base.get_string_buf
...
+    ngx_lua_ffi_cert_pem_to_der = C.ngx_stream_lua_ffi_cert_pem_to_der
+    ngx_lua_ffi_priv_key_pem_to_der = C.ngx_stream_lua_ffi_priv_key_pem_to_der
+    ngx_lua_ffi_ssl_get_tls1_version = C.ngx_stream_lua_ffi_ssl_get_tls1_version
+    ngx_lua_ffi_parse_pem_cert = C.ngx_stream_lua_ffi_parse_pem_cert
+    ngx_lua_ffi_parse_pem_priv_key = C.ngx_stream_lua_ffi_parse_pem_priv_key
+    ngx_lua_ffi_set_cert = C.ngx_stream_lua_ffi_set_cert
+    ngx_lua_ffi_set_priv_key = C.ngx_stream_lua_ffi_set_priv_key
+    ngx_lua_ffi_free_cert = C.ngx_stream_lua_ffi_free_cert
+    ngx_lua_ffi_free_priv_key = C.ngx_stream_lua_ffi_free_priv_key
+end
```
Comment (thibaultcha) [inline lib/ngx/ssl.lua]: style: there used to be 2 blank lines below this, and now a single one only

File: lib/ngx/ssl.lua [hunk 2]
```diff
@@ -93,7 +189,7 @@ function _M.set_der_cert(data)
         error("no request found")
     end
 
-    local rc = C.ngx_http_lua_ffi_ssl_set_der_certificate(r, data, #data,
+    local rc = ngx_lua_ffi_ssl_set_der_certificate(r, data, #data,
```
Comment (thibaultcha) [inline lib/ngx/ssl.lua]: style: the alignment of the arguments is off. Also, maybe now they could all fit on a single line? There are many places in this PR where this is the case as well

File: t/stream/ssl.t [hunk 3]
```diff
@@ -0,0 +1,1826 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
+
+use Test::Nginx::Socket::Lua::Stream;use Cwd qw(cwd);
```
Comment (thibaultcha) [inline t/stream/ssl.t]: Is it intentional that there isn't a new line between those two statements? The `t/ssl.t` test suite has such a new line.

File: t/stream/ssl.t [hunk 4]
```diff
@@ -0,0 +1,1826 @@
...
+
+our $CWD = cwd();
+
+no_long_string();
+#no_diff();
+
+env_to_nginx("PATH=" . $ENV{'PATH'});
+$ENV{TEST_NGINX_LUA_PACKAGE_PATH} = "$::CWD/lib/?.lua;;";
+$ENV{TEST_NGINX_HTML_DIR} ||= html_dir();
+$ENV{TEST_NGINX_SERVER_SSL_PORT} ||= 4443;
```
Comment (thibaultcha) [inline t/stream/ssl.t]: If we run the test suite concurrently, there is a chance that this suite conflicts with `t/ssl.t` (which uses the same env variable and defaults to the same port). Another approach would be to use a unix socket maybe?
```
listen unix:$ENV{TEST_NGINX_HTML_DIR}/nginx-ssl.sock ssl;
```

File: t/stream/ssl.t [hunk 5]
```diff
@@ -0,0 +1,1826 @@
...
+run_tests();
+
+__DATA__
+
+=== TEST 1: clear certs
+--- stream_config
+    lua_package_path "$TEST_NGINX_LUA_PACKAGE_PATH";
+
+    server {
+        listen unix:$TEST_NGINX_HTML_DIR/nginx.sock ssl;
```
Comment (thibaultcha) [inline t/stream/ssl.t]: Oh, this is my previous point. Maybe we should update all `listen` directives for mock upstreams to use this.

File: t/stream/ssl.t (hunk 3 above)
Comment (dndx) [reply t/stream/ssl.t]: No that was definitely not intentional. Fixed.

File: lib/ngx/ssl.lua (hunk 1 above)
Comment (dndx) [reply lib/ngx/ssl.lua]: Fixed.

File: t/stream/ssl.t (hunk 4 above)
Comment (dndx) [reply t/stream/ssl.t]: That's a good idea, although does not really belongs to this PR. Maybe we can have a separate issue for it.

Comment (agentzh) [reply t/stream/ssl.t]: I agree.

Comment (agentzh) [reply t/stream/ssl.t]: @dndx It is a new issue introduced by this PR, so it's better to be addressed in this PR too? AFAIK, the current master does support running tests in parallel.

File: t/stream/ssl.t [hunk 6]
```diff
@@ -19,7 +19,7 @@ no_long_string();
 env_to_nginx("PATH=" . $ENV{'PATH'});
 $ENV{TEST_NGINX_LUA_PACKAGE_PATH} = "$::CWD/lib/?.lua;;";
 $ENV{TEST_NGINX_HTML_DIR} ||= html_dir();
-$ENV{TEST_NGINX_SERVER_SSL_PORT} ||= 4443;
+$ENV{TEST_NGINX_SERVER_SSL_PORT} ||= 4433;
```
Comment (thibaultcha) [inline t/stream/ssl.t]: That works too!

**Pull request merged**